charmcraft/commands/build.py

```diff
@@ -0,0 +1,236 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+
+import logging
+import os
+import pathlib
+import shutil
+import stat
+import subprocess
+import zipfile
+
+from charmcraft.cmdbase import BaseCommand, CommandError
+
+logger = logging.getLogger(__name__)
+
+# Some constants that are used through the code.
+CHARM_METADATA = 'metadata.yaml'
+BUILD_DIRNAME = 'build'
+DISPATCH_FILENAME = 'dispatch'
+
+# The template for the dispatch script
+DISPATCH = """#!/bin/sh
+
+PYTHONPATH=lib {entrypoint_relative_path}
+"""
+
+# The minimum set of hooks to be provided for compatibility with old Juju
+HOOK_NAMES = [
+    'install',
+    'start',
+    'upgrade-charm',
+]
+
+
+def polite_exec(cmd):
+    """Execute a command, only showing output if error."""
+    p = subprocess.Popen(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
+
+    for line in p.stdout:
+        logger.debug(":: %s", line.rstrip())
+    retcode = p.wait()
+
+    if retcode:
+        logger.error("Execution ended in %d for cmd %s", retcode, cmd)
+    return retcode
+
+
+def link(srcpath, destdir):
```

Comment:
Nothing is tested! We need to agree that all this makes sense before jumping into that :)

BTW, maybe we need to make all absolute paths when we read from the cmdline? it would make everything way more explicit in debug logs and everything...